linera-chain/src/test/http_server.rs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A simple HTTP server to use for testing.
+
+use std::{future::IntoFuture, net::Ipv4Addr};
+
+use axum::Router;
+use futures::FutureExt as _;
+use tokio::{net::TcpListener, sync::oneshot};
+
+/// A handle to a running HTTP server.
+///
+/// The server is gracefully shutdown when this handle is dropped.
+pub struct HttpServer {
+    port: u16,
+    _shutdown_sender: oneshot::Sender<()>,
+}
+
+impl HttpServer {
+    /// Spawns a task with an HTTP server serving the routes defined by the [`Router`].
+    ///
+    /// Returns a [`HttpServer`] handle to keep the server running in the background.
+    pub async fn start(router: Router) -> anyhow::Result<Self> {
+        let (shutdown_sender, shutdown_receiver) = oneshot::channel();
+        let shutdown_signal = shutdown_receiver.map(|_| ());
```

Comment:
Isn't the type already `()`? Why the `map`?

Reply:
The `oneshot::Receiver<()>` [implements](https://docs.rs/oneshot/latest/oneshot/struct.Receiver.html#impl-Future-for-Receiver%3CT%3E) `Future<Output = Result<(), oneshot::RecvError>>`, so this just treats the error caused by a dropped sender as the shutdown signal as well.